### DISCUSSION TITLE
v4.9.0-rc.3

### DISCUSSION THREAD

**Discussion by jsjoeio** [coder/code-server#5829] [Announcements]
Code v1.73.1

This is the third pre-release for v4.9.0. 

To install via the install script, pass the `--edge` flag. 

<hr /><em>This discussion was created from the release <a href='https://github.com/coder/code-server/releases/tag/v4.9.0-rc.3'>v4.9.0-rc.3</a>.</em>

**Comment by jsjoeio** [comment]
@benz0li do you mind testing to see if [this issue](https://github.com/coder/code-server/discussions/5824#discussioncomment-4280065) is fixed?

**Reply by benz0li** [reply]
The issue remains:

```bash
$ code-server --version
4.9.0-rc.3 4db51ad34d4a3f7c7e1807b7f1c6b9d062139677 with Code 1.73.1
```

```bash
$ ls -al /opt/code-server/lib/vscode
total 168
drwxr-xr-x   7 root root  4096 Dec  1 21:23 .
drwxr-xr-x   3 root root  4096 Dec  1 21:23 ..
drwxr-xr-x   4 root root  4096 Dec  1 18:28 bin
drwxr-xr-x 106 root root  4096 Dec  1 22:42 extensions
drwxr-xr-x 107 root root  4096 Nov 30 23:30 node_modules
lrwxrwxrwx   1 root root    12 Dec  1 21:23 node_modules.asar -> node_modules
-rw-r--r--   1 root root 91042 Dec  1 20:30 npm-shrinkwrap.json
drwxr-xr-x   3 root root  4096 Dec  1 18:28 out
-rw-r--r--   1 root root  1222 Dec  1 20:30 package.json
-rw-------   1 root root  3813 Dec  1 21:19 product.json
drwxr-xr-x   3 root root  4096 Dec  1 18:28 resources
-rw-r--r--   1 root root 38322 Dec  1 20:30 yarn.lock
```

**Reply by jsjoeio** [reply]
I'm stumped on this one. Thank you for the quick response. I'm going to continue digging in.

**Reply by jsjoeio** [reply]
# Investigation Notes

I'll update this comment as I go.

## What's the problem? 

In the latest release candidate -- 4.9.0-rc.3 -- there's an issue with file permissions on the `lib/vscode/product.json` file and I'm not sure why.

## Okay. How can I see this problem?

1. Visit the [latest release](https://github.com/coder/code-server/releases/tag/v4.9.0-rc.3)
2. Download the `linux-amd64` release
3. Unzip and `cd code-server-4.9.0-rc.3-linux-amd64`
4. Run `ls -al lib/vscode`

**Expected** 
`product.json` should have `644` permissions instead of `600``

**Actual**

It doesn't...
```console
-rw-------    1 jp  staff   3813 Dec  1 13:19 product.json
```

## Hmm...does this issue exist in the other binaries? i.e. the macOS one?

Checking using the same steps as above, yes, that binary has the issue too.

```console
-rw-------    1 jp  staff   3813 Dec  1 13:19 product.json
```

## Well, that's good I guess. Does this issue exist in the npm artifact generated during CI?

Let's check.
1. Download artifact from [here](https://github.com/coder/code-server/actions/runs/3595876003#artifacts)
2. Unzip
3. Run `ls -al lib/vscode`

AH! Check this out:
```console
-rwxr-xr-x   1 jp  staff   3701 Dec  1 10:20 product.json
```

That looks right to me. So something is happening when we go from npm artifact -> native binaries 🤔 

## Woohoo! 🙌🏼 Progress. Tell me about that process. How do you go from npm artifact to native binaries for a release?

Here's an example job in CI where that happens: https://github.com/coder/code-server/actions/runs/3596222251/jobs/6056654392

Basically we do this:
1. download npm artifact (after modifying `package.json` version)
2. build a standalone release
3. upload to release assets

I'm guessing we need to dig into the Step 2.

## Okay. Walk me through it. What happens?

Sure, okay here's what happens:
1. first we set up the CI job
2. we initialize the containers. 
3. we checkout the repo
4. we install node
5. we install the development tools
6. we install yarn
7. we download and decompress the npm artifact (at this point, we know `product.json` has the correct permissions)
8. we build the standalone release
9. we do some other stuff...
10. later we build the packages with `nfpm` which i haven't looked into as much.
11. we upload the assets to the release draft

really, I think we need to dig into Steps 8 and 10. guessing there's something funky happening there.

## Agreed. Let's start with 8. What does "Build standalone release" involve?

I'm going to pull up the [script](https://github.com/coder/code-server/blob/main/ci/build/build-standalone-release.sh) and go through it line by line. 

We source our `./ci/lib.sh` to get some helper functions.

We then use `rsync` to copy release path to the standalone release. We grab the path to node then do a couple `mkdir` and `rsync` more stuff.

We then run `npm install` in the release path and that's it.

It outputs `release-standalone` at the root directory. So if I use `ls -al release-standalone/lib/vscode` I get this:

```console
-rw-r--r--    1 jp  staff   3701 Nov 14 11:29 product.json
```

Therefore, it seems like this script is fine.

## Cool! Let's take a look at that `nfpm` step you mentioned.

So that's going to run the `package` script which runs this script: https://github.com/coder/code-server/blob/main/ci/build/build-packages.sh

I'm mentally drained so let's pick back up tomorrow?

## Sounds like a plan!

Alright I'm back.

## Happy Friday! Ready to do some digging?

Oh yeah.

## Where'd we end?

Looking at the `build-packages.sh` script which will build a package based on your current OS (mac in my case). So let's run `yarn package` and then inspect the results.

So here's what I'm doing:

1. VERSION="1.2.3" yarn package
2. `cd release-packages`
3. `tar -xzf code-server*`
4. `cd code-server*`
5. `ls -al lib/vscode`

And if you look, the file has the right permissions so no issue 🤔

```console
-rw-r--r--    1 jp  staff   3701 Nov 14 11:29 product.json
```

Stumped again...

Next thing I'm thinking is to add a "debug step" to our `release.yaml` for the Linux and macOS steps to run `ls -al lib/vscode` in CI and see if we get the same results as local.

## Good thinking. Let's do it!

Look at this: https://github.com/coder/code-server/actions/runs/3603726145/jobs/6072280970#step:13:21

So running all the same steps in CI produces a different result...After the `yarn package` step, the file permissions are wrong. It makes me wonder if they're correct earlier in the chain 🤔 I'm going to add a couple more debug steps.

**Reply by jsjoeio** [reply]
## Interesting finding?

YES! Check this out. So each platform step (Linux, macOS) downloads the npm package from the workflow after the version has been modified.

When it downloads it, it decompresses it and get this: the file permissions are wrong there!

https://github.com/coder/code-server/actions/runs/3603837519/jobs/6072521305#step:7:14

So now I need to check the file permissions in the "Modify package.json" step.

**Reply by jsjoeio** [reply]
## CHECK THIS OUT

I didn't even realize but look at this block of code: 

https://github.com/coder/code-server/blob/main/.github/workflows/release.yaml#L315-L324

I bet you that's it. My hunch says before this step, file permissions are correct. Then we use `mv` and `jq` and I bet it messes it up! Let's debug and find out.

**Reply by jsjoeio** [reply]
## THAT WAS IT!

After we modify the version in the `package.json` the file permissions change, causing the bug.

![file-permissions](https://user-images.githubusercontent.com/3806031/205355378-cc80b287-44b9-4af1-9c4a-c7b2f4d713a4.png)

https://github.com/coder/code-server/actions/runs/3603913725/jobs/6072677160#step:4:14

Now I need to see if there's a way to `mv` but keep same file permissions.

**Reply by jsjoeio** [reply]
Alright tested locally and I think this should fix it: https://github.com/coder/code-server/pull/5772/commits/0f66360646b842a620af2d5e7ce40da7f76b5931

Going to do another release candidate.

**Reply by benz0li** [reply]
@jsjoeio Thank you for the detailed problem analysis. It's awesome how you do this!

**Reply by jsjoeio** [reply]
Thanks for saying that! It's definitely slower than other approaches, but leaves a note trail for others to see and helps understand everything much better.
